src/main/scala/com/wavesplatform/consensus/PoSCalculator.scala

```diff
@@ -0,0 +1,140 @@
+package com.wavesplatform.consensus
+
+import com.wavesplatform.crypto
+import com.wavesplatform.features.BlockchainFeatures
+import com.wavesplatform.features.FeatureProvider._
+import com.wavesplatform.state.Blockchain
+
+trait PoSCalculator {
+  protected val HitSize: Int              = 8
+  protected val MeanCalculationDepth: Int = 3
+
+  // Min BaseTarget value is 9 because only in this case it is possible to get to next integer value (10)
+  // then increasing base target by 11% and casting it to Long afterward (see lines 55 and 59)
+  protected val MinBaseTarget: Long = 9
+
+  protected val MinBlockDelaySeconds = 53
+  protected val MaxBlockDelaySeconds = 67
+  protected val BaseTargetGamma      = 64
+
+  def generatorSignature(signature: Array[Byte], publicKey: Array[Byte]): Array[Byte] = {
+    val s = new Array[Byte](crypto.DigestSize * 2)
+    System.arraycopy(signature, 0, s, 0, crypto.DigestSize)
+    System.arraycopy(publicKey, 0, s, crypto.DigestSize, crypto.DigestSize)
+    crypto.fastHash(s)
+  }
+
+  def baseTarget(targetBlockDelaySeconds: Long,
+                 prevHeight: Int,
+                 prevBaseTarget: Long,
+                 parentTimestamp: Long,
+                 maybeGreatGrandParentTimestamp: Option[Long],
+                 timestamp: Long): Long
+
+  protected def hit(generatorSignature: Array[Byte]): BigInt = BigInt(1, generatorSignature.take(HitSize).reverse)
+
+  def calculateDelay(hit: BigInt, bt: Long, balance: Long): Long
+
+  def validBlockDelay(genSig: Array[Byte], baseTarget: Long, balance: Long): Long = {
+    calculateDelay(hit(genSig), baseTarget, balance)
+  }
+
+  def validBlockDelay(genSig: Array[Byte], publicKey: Array[Byte], baseTarget: Long, balance: Long): Long = {
+    calculateDelay(hit(generatorSignature(genSig, publicKey)), baseTarget, balance)
+  }
+
+  protected def normalize(value: Long, targetBlockDelaySeconds: Long): Double =
+    value * targetBlockDelaySeconds / (60: Double)
+
+  protected def normalizeBaseTarget(baseTarget: Long, targetBlockDelaySeconds: Long): Long = {
+    val maxBaseTarget = Long.MaxValue / targetBlockDelaySeconds
+    if (baseTarget < MinBaseTarget) MinBaseTarget else if (baseTarget > maxBaseTarget) maxBaseTarget else baseTarget
```

Comment:
Small tip: it could be simpler
```scala
else math.min(maxBaseTarget, baseTarget)
```
or even simpler, but please doesn't change, it's ok

Reply:
`baseTarget.max(MinBaseTarget).min(maxBaseTarget)`